src/NuGet.Core/NuGet.Common/MsBuildStringUtility.cs

```diff
@@ -89,28 +91,51 @@ public static bool IsTrueOrEmpty(string? value)
         /// </summary>
         /// <param name="s">A comma or semicolon delimited list of NuGet log codes.</param>
         /// <returns>An <see cref="IList{T}" /> containing the <see cref="NuGetLogCode" /> values that were successfully parsed from the specified string.</returns>
-        public static IList<NuGetLogCode> GetNuGetLogCodes(string s)
+        public static ImmutableArray<NuGetLogCode> GetNuGetLogCodes(string s)
         {
             // The Split() method already checks for an empty string and returns Array.Empty<string>().
             string[] split = MSBuildStringUtility.Split(s, ';', ',');
 
             if (split.Length == 0)
             {
-                return Array.Empty<NuGetLogCode>();
+                return [];
             }
 
-            List<NuGetLogCode> logCodes = new List<NuGetLogCode>(capacity: split.Length);
+            NuGetLogCode[]? logCodes = null;
+            int index = 0;
 
             for (int i = 0; i < split.Length; i++)
             {
                 if (split[i].StartsWith("NU", StringComparison.OrdinalIgnoreCase) &&
                     Enum.TryParse(value: split[i], ignoreCase: true, out NuGetLogCode logCode))
                 {
-                    logCodes.Add(logCode);
+                    if (logCodes == null)
+                    {
+                        logCodes = ArrayPool<NuGetLogCode>.Shared.Rent(split.Length);
+                    }
+                    else if (logCodes.Length == index)
+                    {
+                        var oldItems = logCodes;
+
+                        logCodes = ArrayPool<NuGetLogCode>.Shared.Rent(logCodes.Length * 2);
```

Comment:
how can this ever happen? 

Wouldn't array pool return an array of at least that size?

Reply:
Yeah, you're right, this can be simplified.